In [1]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 3: No Matter How You Slice It ---</h2><p>The Elves managed to locate the chimney-squeeze prototype fabric for Santa's suit (thanks to <span title="WAS IT YOU">someone</span> who helpfully wrote its box IDs on the wall of the warehouse in the middle of the night).  Unfortunately, anomalies are still affecting them - nobody can even agree on how to <em>cut</em> the fabric.</p>
<p>The whole piece of fabric they're working on is a very large square - at least <code>1000</code> inches on each side.</p>
<p>Each Elf has made a <em>claim</em> about which area of fabric would be ideal for Santa's suit.  All claims have an ID and consist of a single rectangle with edges parallel to the edges of the fabric.  Each claim's rectangle is defined as follows:</p>
<ul>
<li>The number of inches between the left edge of the fabric and the left edge of the rectangle.</li>
<li>The number of inches between the top edge of the fabric and the top edge of the rectangle.</li>
<li>The width of the rectangle in inches.</li>
<li>The height of the rectangle in inches.</li>
</ul>
<p>A claim like <code>#123 @ 3,2: 5x4</code> means that claim ID <code>123</code> specifies a rectangle <code>3</code> inches from the left edge, <code>2</code> inches from the top edge, <code>5</code> inches wide, and <code>4</code> inches tall. Visually, it claims the square inches of fabric represented by <code>#</code> (and ignores the square inches of fabric represented by <code>.</code>) in the diagram below:</p>
<pre><code>...........
...........
...#####...
...#####...
...#####...
...#####...
...........
...........
...........
</code></pre>
<p>The problem is that many of the claims <em>overlap</em>, causing two or more claims to cover part of the same areas.  For example, consider the following claims:</p>
<pre><code>#1 @ 1,3: 4x4
#2 @ 3,1: 4x4
#3 @ 5,5: 2x2
</code></pre>
<p>Visually, these claim the following areas:</p>
<pre><code>........
...2222.
...2222.
.11XX22.
.11XX22.
.111133.
.111133.
........
</code></pre>
<p>The four square inches marked with <code>X</code> are claimed by <em>both <code>1</code> and <code>2</code></em>. (Claim <code>3</code>, while adjacent to the others, does not overlap either of them.)</p>
<p>If the Elves all proceed with their own plans, none of them will have enough fabric. <em>How many square inches of fabric are within two or more claims?</em></p>
</article>


In [2]:
from re import findall


claims_1 = "#123 @ 3,2: 5x4"
claims_2 = """
#1 @ 1,3: 4x4
#2 @ 3,1: 4x4
#3 @ 5,5: 2x2
"""


def overlapping_square_inchis(claims: str) -> int:
    inches = defaultdict(int)
    for claim in claims.strip().splitlines():
        _, left, top, width, height = map(int, findall(r"\d+", claim))
        for w, h in product(range(width), range(height)):
            inches[(left + w, top + h)] += 1
    return sum(1 for c in inches.values() if c > 1)


print(f"Example 1: {overlapping_square_inchis(claims_1)} should be 0")
print(f"Example 2: {overlapping_square_inchis(claims_2)} should be 4")

Example 1: 0 should be 0
Example 2: 4 should be 4


In [3]:
with open("../input/day3.txt") as f:
    puzzle = f.read()

print(f"Part I: {overlapping_square_inchis(puzzle)}")

Part I: 115304


<link href="style.css" rel="stylesheet"></link>
...<main>

<p>Your puzzle answer was <code>115304</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>Amidst the chaos, you notice that exactly one claim doesn't overlap by even a single square inch of fabric with any other claim. If you can somehow draw attention to it, maybe the Elves will be able to make Santa's suit after all!</p>
<p>For example, in the claims above, only claim <code>3</code> is intact after all claims are made.</p>
<p><em>What is the ID of the only claim that doesn't overlap?</em></p>
</article>

</main>


In [4]:
from more_itertools import first


def claim_without_overlap(claims: str) -> int:
    inches = defaultdict(list[int])
    ids = set()

    for claim in claims.strip().splitlines():
        id, left, top, width, height = map(int, findall(r"\d+", claim))
        for w, h in product(range(width), range(height)):
            inches[(left + w, top + h)].append(id)
            ids.add(id)

    ids_overlap = {v for vs in inches.values() for v in vs if len(vs) > 1}
    return first(ids - ids_overlap)


print(f"Example 1: {claim_without_overlap(claims_1)} should be 123")
print(f"Example 2: {claim_without_overlap(claims_2)} should be 3")

Example 1: 123 should be 123
Example 2: 3 should be 3


In [5]:
print(f"Part II: {claim_without_overlap(puzzle)}")

Part II: 275


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>275</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
